src/System.Runtime/tests/System/Reflection/PointerTests.cs

```diff
@@ -0,0 +1,197 @@
+        }
+
+        [Fact]
+        public void Box_NonPointerType()
+        {
+            Assert.Throws<ArgumentException>(() =>
+            {
+                Pointer.Box((void*)0, typeof(int));
+            });
+        }
+
+        [Fact]
+        public void Unbox_Null()
+        {
+            Assert.Throws<ArgumentException>(() =>
+            {
+                Pointer.Unbox(null);
+            });
+        }
+
+        [Fact]
+        public void Unbox_NotPointer()
+        {
+            Assert.Throws<ArgumentException>(() =>
+            {
+                Pointer.Unbox(new object());
+            });
+        }
+
+        [Theory]
+        [MemberData(nameof(Pointers))]
+        public void PointerValueRoundtrips(int value)
+        {
+            void* ptr = (void*)value;
+            void* result = Pointer.Unbox(Pointer.Box(ptr, typeof(int*)));
+            Assert.Equal((IntPtr)ptr, (IntPtr)result);
+        }
+
+        public static IEnumerable<object[]> Pointers =>
+            new[]
+            {
+                new object[] { 0 },
+                new object[] { 1 },
+                new object[] { -1 },
+                new object[] { int.MaxValue },
+                new object[] { int.MinValue },
+            };
+
+        [Theory]
+        [MemberData(nameof(Pointers))]
+        public void PointerFieldSetValue(int value)
+        {
+            var obj = new PointerHolder();
+            FieldInfo field = typeof(PointerHolder).GetField("field");
+            field.SetValue(obj, Pointer.Box((void*)value, typeof(int*)));
+            Assert.Equal(value, (int)obj.field);
+        }
+
+        [Theory]
+        [MemberData(nameof(Pointers))]
+        public void PointerFieldSetValue_InvalidType(int value)
+        {
+            var obj = new PointerHolder();
+            FieldInfo field = typeof(PointerHolder).GetField("field");
+            Assert.Throws<ArgumentException>(() =>
+            {
+                field.SetValue(obj, Pointer.Box((void*)value, typeof(long*)));
+            });
+        }
+
+        [Theory]
+        [MemberData(nameof(Pointers))]
+        public void PointerFieldGetValue(int value)
+        {
+            var obj = new PointerHolder();
+            obj.field = (int*)value;
+            FieldInfo field = typeof(PointerHolder).GetField("field");
+            object actualValue = field.GetValue(obj);
+            Assert.IsType<Pointer>(actualValue);
+            void* actualPointer = Pointer.Unbox(actualValue);
+            Assert.Equal(value, (int)actualPointer);
+        }
+
+        [Theory]
+        [MemberData(nameof(Pointers))]
+        public void PointerPropertySetValue(int value)
+        {
+            var obj = new PointerHolder();
+            PropertyInfo property = typeof(PointerHolder).GetProperty("Property");
+            property.SetValue(obj, Pointer.Box((void*)value, typeof(char*)));
+            Assert.Equal(value, (int)obj.Property);
+        }
```

Comment:
Please test that passing the parameter in as a boxed IntPtr does something. (I believe that's also supposed to work)